jb/fftw/time_delay_estimator_many.hpp

```diff
@@ -0,0 +1,185 @@
+
+  /**
+   * @name Type traits
+   */
+  /// The input timeseries type
+  using array_type = array_t;
+
+  /// The values stored in the input timeseries
+  using element_type =
+      typename jb::detail::array_traits<array_type>::element_type;
+
+  /// Extract T out of std::complex<T>, otherwise simply T.
+  using precision_type =
+      typename jb::extract_value_type<element_type>::precision;
+
+  /// complex_type use by FFT plans frequency array type
+  using complex_type = std::complex<precision_type>;
+
+  /// The type used to store the DFT of the input timeseries
+  using frequency_array_type =
+      typename jb::detail::aligned_container<complex_type,
+                                             array_type>::array_type;
+
+  /// The type used to store the inverse of the DFT
+  using output_array_type =
+      typename jb::detail::aligned_container<precision_type,
+                                             array_type>::array_type;
+
+  /// The execution plan to apply the (forward) DFT
+  using dplan = jb::fftw::plan<array_type, frequency_array_type>;
+
+  /// The execution plan to apply the inverse (aka backward) DFT
+  using iplan = jb::fftw::plan<frequency_array_type, output_array_type>;
+
+  /// The type used to store the TDE confidence between two timeseries
+  using confidence_type = jb::fftw::tde_result<array_t, precision_type>;
+
+  /// The type used to store the TDE argmax between two timeseries
+  using argmax_type = jb::fftw::tde_result<array_t, std::size_t>;
+
+  /// The sqr sum of a timeseries
+  using sum2_type = jb::fftw::tde_result<array_t, precision_type>;
+
+  //@}
+
+  /**
+   * Constructs a time delay estimator using @a a and @a b as prototypes
+   * for the arguments.
+   *
+   * The optimal algorithm to compute the  FFTs used in the cross correlation
+   * depends on the size of the input parameters and their memory alignment.
+   *
+   * The FFTW library casts the constness of their arguments to change their
+   * values, so we are making it explicit pasing non-const reference arguments
+   * to the constructor.
+   *
+   * @param a multi array timeseries
+   * @param b multi array timeseries
+   */
+  time_delay_estimator_many(array_type& a, array_type& b)
+      : tmpa_(jb::detail::array_shape(a))
+      , tmpb_(jb::detail::array_shape(b))
+      , a2tmpa_(create_forward_plan(a, tmpa_, planning_flags()))
+      , b2tmpb_(create_forward_plan(b, tmpb_, planning_flags()))
+      , out_(jb::detail::array_shape(a))
+      , tmpa2out_(create_backward_plan(tmpa_, out_, planning_flags()))
+      , nsamples_(jb::detail::nsamples(a))
+      , num_timeseries_(jb::detail::element_count(a) / nsamples_) {
+    if (a.size() != b.size()) {
+      throw std::invalid_argument("size mismatch in time_delay_estimator ctor");
+    }
+  }
+
+  /**
+   * Compute the time-delay estimate between two timeseries a and b.
+   *
+   * @param confidence to return the confidence of TDE(a,b)
+   * @param argmax to return argmax of TDE(a,b)
```

Comment:
Actually, TDE(a,b) **is** argmax(Cross-correlation(a, b)), so a better name might be **result** instead of *argmax*.